### PR TITLE
Entity Timeline

### LINKED ISSUE #54: Entity Timeline
## Why
The conceptual data model of geovistory, based on CIDOC-CRM, is event centered: The persistent items (persons, places) are related to each other and to literals describing their attributes through time referenced temporal entities (birth, death, occupations, appellations, localizations, etc.). Thus, the events and their time are very central in the data. This issue is about bringing the time referenced data to a visual form. 

The **goal** is to allow readers to quickly see events in chronological order. For example: The events in the life of a person.

### Time in data

Each temporal entity can have a time span. Each time span can have 6 time properties in order to narrow down its localization and duration within a temporal reference system (calendar and time).

| property | label in diagram |
|---|---|
| [crm:P82 at some time within](https://ontome.net/property/72/namespace/1) | At some time within |
| [crm:P81 ongoing throughout](https://ontome.net/property/71/namespace/1) | Ongoing throughout |
| [crm:P82a begin of the begin](https://ontome.net/property/152/namespace/1) | Earliest possible begin |
| [crm:P81a end of the begin](https://ontome.net/property/150/namespace/1) | Begin |
| [crm:P81b begin of the end](https://ontome.net/property/151/namespace/1) | End |
| [crm:P82b end of the end](https://ontome.net/property/153/namespace/1) | Latest possible end |

Image

### Time in visualisation

To visualise this kind of data on a timeline, a number of questions has to be answered:

**Do we use an Absolute chronology or Relative chronology?** 
An Absolute chronology is a chronological order which outlines events in terms of their absolute time span. A relative chronology is a chronological order that outlines events relative to other events. It can be used to compare different periods, or to analyze and understand the sequence of events in a specific period. 

**How to deal with the 6 time span properties?** 
The visualization of a time span could display all 6 properties in a visually distinguishable way (as on the diagram above), it could reduce the 6 props to 2 (start and end) or to one (point in time). 

**How to display time spans in a relative chronology?** 
Assumed we have two simplified time spans A and B with start and end, they can have different relationships:
- no overlap: A ends before B starts
- overlap: A starts before B, B starts before A ends, A ends before B ends
- contains: A starts before B and ends after B  
The visualization would either reduce every event to a simple point in time (which is good for short events like a Birth, but bad for longer phases) or it would have to visualize the relationships sketched above.

**Decision**
After the discussion (see below) we create a time line visualization in the style of a Gantt chart, with an absolute chronology referencing the start and end dates of time spans.

### PR DESCRIPTION
**Pull request by perrauda**
Resolves #54

### CODE REVIEW

**Review by joschne — COMMENTED** [review]
Some considerations:
- The height per row should be ~30px
- The labels of the bar should be left aligned and overflow the bar
- The bar should be of lighter color and the label of dark color
- The X-Axis should be fixed on bottom, in case the chart overflows the height. 
- The tooltip of a bar should shows strange numbers
- The tooltip should show {class label} {entity label} and provide a link to open the temporal entity URI in a new browser tab.

**Review by joschne — CHANGES_REQUESTED** [review]
I request changes as commented

File: packages/design-system-web/src/components/geov-timeline-gantt/geov-timeline-gantt.css
Comment (joschne) [inline packages/design-system-web/src/components/geov-timeline-gantt/geov-timeline-gantt.css:1]: I suggest to not set a max-height. This should be responsibility of the container. (Here in PluginTimeline.tsx) I also suggest to set the button to absolute position in top right corner.

:host {
  display: block;
}

.scrollBox {
  overflow-y: auto;
  height: 100%;
}

ion-button {
  position: absolute;
  right: 16px;
  top: 8px;
}

**Comment by perrauda** [comment]
Now everything is almost functional, except for scrolling for large data. The double axis works, but the real axis is not yet hidden.

File: packages/design-system-web/src/components/geov-timeline-gantt/geov-timeline-gantt.tsx
Comment (joschne) [inline packages/design-system-web/src/components/geov-timeline-gantt/geov-timeline-gantt.tsx:1]: I would suggest to disable the zoom wheel in favor of normal scroll behavior.

Instead you could add zoom buttons:

```
// in render()
  <div class="gantt-buttons">
    <ion-button size="small" onClick={() => this.resetZoom()}>
      Reset zoom
    </ion-button>
    <ion-button size="small" onClick={() => this.zoomOut()}>
      -
    </ion-button>
    <ion-button size="small" onClick={() => this.zoomIn()}>
      +
    </ion-button>
  </div>
```

```
// in class
  zoomOut() {
    Chart.getChart(this.el).zoom(0.5, 'zoom');
  }
  zoomIn() {
    Chart.getChart(this.el).zoom(1.333333, 'zoom');
  }
```

**Review by joschne — CHANGES_REQUESTED** [review]

File: packages/design-system-web/src/components/geov-timeline-gantt/geov-timeline-gantt.tsx [hunk 1]
```diff
@@ -0,0 +1,366 @@
...
+                    }
+                  } else if (Array.isArray(dataItem)) {
+                    // For data in the format of [startDate, endDate]
+                    const startDate = new Date(dataItem[0]);
+                    const endDate = new Date(dataItem[1]);
+                    if (startDate.toDateString() === endDate.toDateString()) {
+                      dateLabel = startDate.toISOString().split('T')[0];
+                    } else {
+                      dateLabel = startDate.toISOString().split('T')[0] + ' to ' + endDate.toISOString().split('T')[0];
+                    }
```
[the remark this answers is not in the document]
Comment (perrauda) [reply packages/design-system-web/src/components/geov-timeline-gantt/geov-timeline-gantt.tsx:223]: Done. Removed the unreachable if. Probably to test with manual data and not from sparql.

**Review by joschne — APPROVED** [review]

**Pull request merged**